### PR TITLE
Multihead-attention slim enhance

### PR DESCRIPTION
**Pull request by YIYANGCAI**
## Type of Change

Enhancement

## Description

Enhance multihead attention slim automatic process, make it fit more llms.
Support MHA training-based pruning.

## Expected Behavior & Potential Risk

One general set of codes to perform a wide range of llm models (similar to present FFN2 slim feature)

## How has this PR been tested?

PreCI

## Dependency Change?

None

### CODE REVIEW

**Comment by YIYANGCAI** [comment]
w.i.p, once all types of models are tested, then this pr are ready to be reviewed/merged

**Comment by YIYANGCAI** [comment]
Use a strategy of name matching (no need to specify model types or names) to hook mha attributes. Tested successfully on BERT, OPT, LLAMA-7b, GPTJ-6B

File: neural_compressor/compression/pruner/pruners.py [hunk 1]
```diff
@@ -73,6 +77,10 @@ def get_pruner(config, modules):
     Raises: AssertionError: Cuurently only support pruners that have been registered in PRUNERS.
     """
     ## do the ugly work here
+    ## check if it is doing self-multihead-attention pruning
+    if "mha" in config["pattern"]:
+        return PRUNERS["mha"](config, modules)
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/pruners.py:82]: block_mask pruner could not apply on mha pattern currently, please add a check here

**Review by wenhuach21 — APPROVED** [review]

**Comment by YIYANGCAI** [comment]
UT & README updated. Examples refined. Ready to be merged.

File: neural_compressor/compression/pruner/model_slim/weight_slim.py [hunk 2]
```diff
@@ -271,62 +280,120 @@ def __call__(self, masks=None, round_value=32):
...
+        {
+            'qkv_name': ['query_layer_name', 'key_layer_name', 'value_layer_name'],
+            'ffn_name': ['attention_ffn_name'],
+            'mha_name': ['mha_name'] (keep not change),
+            'qkv_module': [torch.nn.Linear, torch.nn.Linear, torch.nn.Linear],
+            'ffn_module': [torch.nn.Linear],
+            'mha_module': [torch.nn.Module] (keep not change),
+        }   
+        """
+        # import pdb;pdb.set_trace()
```
Comment (WeiweiZhang1) [inline neural_compressor/compression/pruner/model_slim/weight_slim.py:296]: Redundant code

File: neural_compressor/compression/pruner/patterns.py [hunk 3]
```diff
@@ -1326,3 +1328,43 @@ def update_progressive_masks(self, pre_masks, cur_masks, scores, progressive_ste
         # we only have to handle global score or local score
         return ProgressivePatternUtils.update_progressive_masks_scores_order(pre_masks, cur_masks, scores, \
                 progressive_step, progressive_configs)
+
+@register_pattern('MHA')
+class PatternMHA(BasePattern):
+    """Pruning Pattern.
+    
+    A Pattern class derived from Pattern. In this pattern, N out of every M continuous weights will be pruned.
```
Comment (WeiweiZhang1) [inline neural_compressor/compression/pruner/patterns.py:1336]: Please correct the description of the MHA pruner.

File: neural_compressor/compression/pruner/patterns.py [hunk 4]
```diff
@@ -1326,3 +1328,43 @@ def update_progressive_masks(self, pre_masks, cur_masks, scores, progressive_ste
...
+        config: A config dict object that contains the pattern information.
+        
+    Attributes:
+        N: The number of elements to be pruned in a weight sequence.
+        M: The size of the weight sequence.
+    """
+
+    def __init__(self, config, modules = None):
+        self.is_global = config.pruning_scope == "global"
+        # 111
```
Comment (WeiweiZhang1) [inline neural_compressor/compression/pruner/patterns.py:1350]: Meaningless comments that may affect user experience

**Pull request merged**